tests/phpunit/integration/Modules/Analytics_4Test.php

```diff
@@ -1666,6 +1667,262 @@ public function test_get_conversion_events( $access_token ) {
 		$this->assertEquals( "/v1beta/properties/$property_id/conversionEvents", $request_url['path'] );
 	}
 
+	public function test_get_enhanced_mesurement_settings__required_params() {
+		// Grant READONLY_SCOPE so request doesn't fail.
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			array_merge(
+				$this->authentication->get_oauth_client()->get_required_scopes(),
+				(array) Analytics::READONLY_SCOPE
+			)
+		);
+
+		$data = $this->analytics->get_data(
+			'enhanced-measurement-settings',
+			array()
+		);
+
+		// Verify if the propertyID is required.
+		$this->assertWPErrorWithMessage( 'Request parameter is empty: propertyID.', $data );
+		$this->assertEquals( 'missing_required_param', $data->get_error_code() );
+		$this->assertEquals( array( 'status' => 400 ), $data->get_error_data( 'missing_required_param' ) );
+
+		$data = $this->analytics->get_data(
+			'enhanced-measurement-settings',
+			array(
+				'propertyID' => '123456789',
+			)
+		);
+
+		// Verify if the webDataStreamID is required.
+		$this->assertWPErrorWithMessage( 'Request parameter is empty: webDataStreamID.', $data );
+		$this->assertEquals( 'missing_required_param', $data->get_error_code() );
+		$this->assertEquals( array( 'status' => 400 ), $data->get_error_data( 'missing_required_param' ) );
+	}
+
+	public function test_get_enhanced_mesurement_settings() {
+		$property_id        = '123456789';
+		$web_data_stream_id = '654321';
+
+		$this->analytics->get_settings()->merge(
+			array(
+				'propertyID'      => $property_id,
+				'webDataStreamID' => $web_data_stream_id,
+			)
+		);
+
+		// Grant READONLY_SCOPE so request doesn't fail.
+		$this->authentication->get_oauth_client()->set_granted_scopes(
+			array_merge(
+				$this->authentication->get_oauth_client()->get_required_scopes(),
+				(array) Analytics::READONLY_SCOPE
+			)
+		);
+
+		FakeHttp::fake_google_http_handler(
+			$this->analytics->get_client(),
+			$this->create_fake_http_handler( $property_id )
```

Comment:
If you look at the implementation `create_fake_http_handler()`, you'll see that the fake handler it returns doesn't explicitly support the enhanced measurement case (i.e. the `/v1alpha/properties/$property_id/dataStreams/$web_data_stream_id/enhancedMeasurementSettings` path), and is therefore simply defaulting to an empty 200 response for these enhanced measurement test cases.

That means it's not currently possible to add the extra step of asserting that the object returned from the GA4 API is indeed being correctly propagated through to the response here.

Seeing as the path relies on `$web_data_stream_id` as well as `$property_id`, rather than updating `create_fake_http_handler()` which is used elsewhere in tests which don't have `$web_data_stream_id` in scope, I'd instead suggest adding an additional `create_enhanced_measurement_fake_http_handler()` method, which takes the args `$property_id` and `$web_data_stream_id`, and which returns a fake handler that does handle the enhanced measurement path, returning a mock object conforming to the `EnhancedMeasurementSettings` schema.

With this in place, it can be used in the various enhanced measurement test cases you've introduced here, and will allow additional assertions of the response, for a slightly more robust set of tests in keeping with other test cases in this file.

Reply:
Updated ✅